(WIP) Fix gpcheckcat failure caused by duplicated oids.

macroyuyang opened the pull request:
If target relation has oid, we reconstruct the tuple slot so that new oid could generated for final tuple inserted. In this fix, we just try to get failure test case pass so we don't dive into the plan change why mapping or junk filter has been removed which causes scribbling problem by operating on original physical tuple. We run it locally for the ICW (orca & without orca). The failures seem not related to this fix. Better solution is welcomed for this fix.

Signed-off-by: Xiaoran Wang <xiwang@pivotal.io>

macroyuyang committed:
Fix gpcheckcat failure caused by duplicated oids. We just restruct partslot
so that new oid could be generated for it.

Signed-off-by: Xiaoran Wang <xiwang@pivotal.io>

hlinnaka commented:
On 25 October 2017 11:49:51 CEST, Yu Yang <notifications@github.com> wrote:
>If target relation has oid, we reconstruct the tuple slot so that new
>oid could generated for final tuple inserted. In this fix, we just try
>to get failure test case pass so we don't dive into the plan change why
>mapping or junk filter has been removed which causes scribbling problem
>by operating on original physical tuple. We run it locally for the ICW
>(orca & without orca). The failures seem not related to this fix.
>Better solution is welcomed for this fix.
>
>Signed-off-by: Xiaoran Wang <xiwang@pivotal.io>
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/greenplum-db/gpdb-postgres-merge/pull/3
>
>-- Commit Summary --
>
>* Fix gpcheckcat failure caused by duplicated oids. We just restruct
>partslot
>
>-- File Changes --
>
>    M src/backend/executor/execDML.c (22)
>    M src/backend/executor/execMain.c (220)
>
>-- Patch Links --
>
>https://github.com/greenplum-db/gpdb-postgres-merge/pull/3.patch
>https://github.com/greenplum-db/gpdb-postgres-merge/pull/3.diff

Can you do the whitespace fixes as a separate commit or PR, directly to master, please? Makes it difficult to see what's actually changed here.

This seems to revert the recent changes to ExecInsert, to copy the tuple. I don't understand why.

What is the test case that this fixes? Was it in the gpcheckcat test? Should we add a more easy-to-run test to src/test/regress for this?

- Heikki

- Heikki

macroyuyang committed:
Fix extra whitespace in execMain.c

Signed-off-by: Xiaoran Wang <xiwang@pivotal.io>

macroyuyang commented:
Split into two commits (one for whitespace fix and one for logic change). This fix following cases in MM_gpcheckcat:
set allow_system_table_mods=DML;
update pg_index set indisunique='f' where indexrelid=(select oid from pg_class where relname='pg_compression_compname_index');
insert into pg_compression select * from pg_compression limit 1;
update pg_index set indisunique='t' where indexrelid=(select oid from pg_class where relname='pg_compression_compname_index');
select compname, count(*) from pg_compression group by compname having count(*) > 1;

Before fix, the statement "insert into pg_compression select * from pg_compression limit 1;" will errors out "duplicate oids", then case failed. Because we don't change oid passed by physical tuple, it will error out when inserting indexes. In this fix, we just reconstruct the tuple slot, and the HeapTuple is already a copy instead of original physical disk one, so I think we could change ExecCopySlotHeapTuple to ExecFetchSlotHeapTuple. It is good idea I could add the case in ICW, but even this test change the catalog table.

macroyuyang commented:
Thanks for the reviewing. We also tried another way to make a copied HeapTuple inside partslot, and then call HeapTupleSetOid(TupGetHeapTuple(partslot), InvalidOid) if the target relation hasoids, then let heap_insert to generate new oid for it. But it is tricky.

jchampio commented:
> This seems to revert the recent changes to ExecInsert, to copy the tuple. I don't understand why.

Unfortunately the new tests don't fail on my machine if I revert the tuple copy, which I think explains why this seemed to "pass" for Max. I'll try to write a test that does fail.

jchampio commented:
@macroyuyang Commit 515bd50f now has a test that fails with this patch. It's messy to read, but hopefully it clarifies a little bit about why we cannot get away with not copying the tuple.

macroyuyang commented:
@jchampio I just pulled -r branch iteration_3_redux and rebuild, but when I execute following SQL:
set allow_system_table_mods=DML;
update pg_index set indisunique='f' where indexrelid=(select oid from pg_class where relname='pg_compression_compname_index');
insert into pg_compression select * from pg_compression limit 1;

Error still happens:
est=# set allow_system_table_mods=DML;
SET
test=# update pg_index set indisunique='f' where indexrelid=(select oid from pg_class where relname='pg_compression_compname_index');
UPDATE 1
test=# insert into pg_compression select * from pg_compression limit 1;
WARNING:  OID assigned for tuple in unrecognized system catalog "pg_compression"
ERROR:  duplicate key value violates unique constraint "pg_compression_oid_index"
DETAIL:  Key (oid)=(3060) already exists.

Besides, I will test your commit [515bd50](https://github.com/greenplum-db/gpdb-postgres-merge/commit/515bd50f1110e39df1217dcd61559d0d94bdc38fl)

macroyuyang commented:
@jchampio I am curious why there is no error when I running your pg_regress test:
test=# create external web table cmd(a text)
  execute E'PGOPTIONS="-c gp_session_role=utility"                         \\
    psql -p $GP_MASTER_PORT $GP_DATABASE $GP_USER -c                       \\
      "CREATE TABLE execinsert_test (id int4, t text);                     \\
       INSERT INTO execinsert_test values (1, ''foo'');"                   \\
                                                                           \\
    && PGOPTIONS="-c gp_session_role=utility"                              \\
    psql -p $GP_MASTER_PORT $GP_DATABASE $GP_USER -c                       \\
      "begin;                                                              \\
       INSERT INTO execinsert_test select * FROM execinsert_test;          \\
       rollback;"                                                          \\
                                                                           \\
    && PGOPTIONS="-c gp_session_role=utility"                              \\
    psql -p $GP_MASTER_PORT $GP_DATABASE $GP_USER test-#   execute E'PGOPTIONS="-c gp_session_role=utility"                         \\
test'#     psql -p $GP_MASTER_PORT $GP_DATABASE $GP_USER -c                       \\
test'#       "CREATE TABLE execinsert_test (id int4, t text);                     \\
test'#        INSERT INTO execinsert_test values (1, ''foo'');"                   \\
test'#                                                                            \\
test'#     && PGOPTIONS="-c gp_session_role=utility"                              \\
test'#     psql -p $GP_MASTER_PORT $GP_DATABASE $GP_USER -c                       \\
test'#       "begin;                                                              \\
test'#        INSERT INTO execinsert_test select * FROM execinsert_test;          \\
test'#        rollback;"                                                          \\
test'#                                                                            \\
test'#     && PGOPTIONS="-c gp_session_role=utility"                              \\
test'#     psql -p $GP_MASTER_PORT $GP_DATABASE $GP_USER -c                       \\
test'#       "select * from execinsert_test;"                                     \\
test'#                                                                            \\
test'#     && PGOPTIONS="-c gp_session_role=utility"                              \\
test'#     psql -p $GP_MASTER_PORT $GP_DATABASE $GP_USER -c                       \\
test'#       "drop table execinsert_test;"'
test-#   on master format 'text';
CREATE EXTERNAL TABLE
test=# select * from cmd;
     a
------------
 INSERT 0 1
 ROLLBACK
  id | t
 ----+---
 (0 rows)

 DROP TABLE
(7 rows)

Did I miss something?

macroyuyang committed:
Fix failure by non-proper has_oids value when executing nodeSplitUpdate for
optimizer=on. The failure could be reproduced by:
set optimizer=on;
DROP TABLE IF EXISTS dml_heap_r;
DROP TABLE IF EXISTS dml_heap_p;

CREATE TABLE dml_heap_r (col1 serial,a int , b int default -1, c text) WITH OIDS DISTRIBUTED BY (a);
CREATE TABLE dml_heap_p (col1 serial, a numeric, b decimal) WITH OIDS DISTRIBUTED BY (a,b);

INSERT INTO dml_heap_p(a,b) SELECT id as a, id as b FROM (SELECT * FROM generate_series(1,2) as id) AS x;
INSERT INTO dml_heap_p(a,b) VALUES(NULL,NULL);

INSERT INTO dml_heap_r(a,b,c) VALUES(generate_series(1,2),generate_series(1,2),'r');
INSERT INTO dml_heap_r(a,b,c) VALUES(NULL,NULL,NULL);

SELECT COUNT(*) FROM dml_heap_r;
SELECT COUNT(*) FROM dml_heap_p;
SELECT SUM(a) FROM dml_heap_r;

DROP TABLE IF EXISTS tempoid;
CREATE TABLE tempoid as SELECT oid,col1,a FROM dml_heap_r ORDER BY 1;

SELECT SUM(dml_heap_r.a) FROM dml_heap_p, dml_heap_r WHERE dml_heap_r.b = dml_heap_p.a;

UPDATE dml_heap_r SET a = dml_heap_r.a FROM dml_heap_p WHERE dml_heap_r.b = dml_heap_p.a;

hlinnaka commented:
This has now been taken care of in commit 69e9504bb918789d06aa2d17aba800fe20d2d6b3. I hope, we'll see if the pipeline agrees ;-)
